guides/keras_cv/simsiam-with-kerascv.py

```diff
@@ -0,0 +1,604 @@
+    as_supervised=True,
+    batch_size=-1,
+)
+x_test, y_test = tf.cast(x_test, tf.float32), tf.cast(y_test, tf.float32)
+
+# Compute the indicies for query, index, val, and train splits
+query_idxs, index_idxs, val_idxs, train_idxs = [], [], [], []
+for cid in range(ds_info.features["label"].num_classes):
+    idxs = tf.random.shuffle(tf.where(y_raw_train == cid))
+    idxs = tf.reshape(idxs, (-1,))
+    query_idxs.extend(idxs[:100])  # 200 query examples per class
+    index_idxs.extend(idxs[100:200])  # 200 index examples per class
+    val_idxs.extend(idxs[200:300])  # 100 validation examples per class
+    train_idxs.extend(idxs[300:])  # The remaining are used for training
+
+random.shuffle(query_idxs)
+random.shuffle(index_idxs)
+random.shuffle(val_idxs)
+random.shuffle(train_idxs)
+
+
+def create_split(idxs: list) -> tuple:
+    x, y = [], []
+    for idx in idxs:
+        x.append(x_raw_train[int(idx)])
+        y.append(y_raw_train[int(idx)])
+    return tf.convert_to_tensor(np.array(x), dtype=tf.float32), tf.convert_to_tensor(
+        np.array(y), dtype=tf.int64
+    )
+
+
+x_query, y_query = create_split(query_idxs)
+x_index, y_index = create_split(index_idxs)
+x_val, y_val = create_split(val_idxs)
+x_train, y_train = create_split(train_idxs)
+
+PRE_TRAIN_STEPS_PER_EPOCH = tf.data.experimental.cardinality(train_ds) // BATCH_SIZE
+PRE_TRAIN_STEPS_PER_EPOCH = int(PRE_TRAIN_STEPS_PER_EPOCH.numpy())
+
+print(
+    tabulate(
+        [
+            ["train", tf.data.experimental.cardinality(train_ds), None],
+            ["val", x_val.shape, y_val.shape],
+            ["query", x_query.shape, y_query.shape],
+            ["index", x_index.shape, y_index.shape],
+            ["test", x_test.shape, y_test.shape],
+        ],
+        headers=["Examples", "Labels"],
+    )
+)
+
+"""
+## Augmentations
+Self-supervised networks require at least two augmented "views" of each example. This can be created using a DataSet and an augmentation function. The DataSet treats each example in the batch as its own class and then the augment function produces two separate views for each example.
+
+This means the resulting batch will yield tuples containing the two views, i.e.,
+Tuple[(BATCH_SIZE, 32, 32, 3), (BATCH_SIZE, 32, 32, 3)].
+
+TensorFlow Similarity provides several random augmentation functions, and here we combine augmenters from the simCLR module to replicate the augmentations used in simsiam.
+"""
+
+"""
+## Augmentation
+
+Now that we have all of our datasets produced, we can move on to dataset augmentation.
+Using KerasCV, it is trivial to construct an augmenter that performs as the one
+described in the original SimSiam paper.  Lets do that below.
+"""
+
+target_size = (96, 96)
+crop_area_factor = (0.08, 1)
+aspect_ratio_factor = (3 / 4, 4 / 3)
+grayscale_rate = 0.2
+color_jitter_rate = 0.8
+brightness_factor = 0.2
+contrast_factor = 0.8
+saturation_factor = (0.3, 0.7)
+hue_factor = 0.2
+
+augmenter = keras_cv.layers.Augmenter(
```

Comment:
at one point I thought this would be good to have a keras_cv API; but honestly just inlining it is very readable.

There is no question as to "What does a siclr augmenter do?"